File: LS-OPT_test/Old/createBC.m
clc; close all; clear all;
%%
readfile = 'lsppout';
writefile = 'bcinfo.k';
%%
ytop =  20; % upper bound geometry
ybot = -20; % lower bound geometry

TOL_g = 1e-4; % geometric tolerance
%% read lsspout file
fileID = fopen(readfile,'r');

tline = fgetl(fileID);

flag1 = 0;                                          % flag for node section
nodesz = [];
nodesy = [];
while ischar(tline)
    % start analyzing after NODE
    if contains(tline,'NODE')
        flag1 = 1;
    end
  
    if flag1
%         disp(tline)
        data = str2num(tline);
        
        if ~isempty(data)
            % store data for BC z
            nodesz = [nodesz; data(1,1)];
            
            % store data for BC y
            if data(1,3) > ytop-TOL_g || data(1,3) < ybot+TOL_g
                nodesy = [nodesy; data(1,1)];
            end
        end
    end
    % Read next line
    tline = fgetl(fileID);
end
fclose(fileID);

%% Write output file
if isfile(writefile)
    delete(writefile)
%     fprintf('file deleted\n');
end

edit(writefile);

% Write Z node list
nodesz = [nodesz; zeros(8-mod(length(nodesz),8),1)];

fid = fopen(writefile,'w');
fprintf(fid,'*KEYWORD\n');
fprintf(fid,'*SET_NODE_LIST\n');
fprintf(fid,'$#     sid       da1       da2       da3       da4    solver\n');
fprintf(fid,'         1       0.0       0.0       0.0       0.0MECH\n');
fprintf(fid,'$#    nid1      nid2      nid3      nid4      nid5      nid6      nid7      nid8\n');

linenr = 0;
for i = 1:length(nodesz)
    linenr = linenr + 1;
    
    if nodesz(i) < 10
        fprintf(fid,'         %d',nodesz(i));
    elseif nodesz(i) >= 10 && nodesz(i) < 100
        fprintf(fid,'        %d',nodesz(i));
    elseif nodesz(i) >= 100 && nodesz(i) < 1000
        fprintf(fid,'       %d',nodesz(i));
    elseif nodesz(i) >=1000 && nodesz(i) < 10000
        fprintf(fid,'      %d', nodesz(i));
    else
        fprintf(fid,'     %d',nodesz(i));
    end
 
    if linenr == 8
        fprintf(fid,'\n');
        linenr = 0;
    end
end

% Write Y node list
nodesy = [nodesy; zeros(8-mod(length(nodesy),8),1)];
fprintf(fid,'*SET_NODE_LIST\n');
fprintf(fid,'         2       0.0       0.0       0.0       0.0MECH\n');

linenr = 0;
for i = 1:length(nodesy)
    linenr = linenr + 1;
    
    if nodesy(i) < 10
        fprintf(fid,'         %d',nodesy(i));
    elseif nodesy(i) >= 10 && nodesy(i) < 100
        fprintf(fid,'        %d',nodesy(i));
    elseif nodesy(i) >= 100 && nodesy(i) < 1000
        fprintf(fid,'       %d',nodesy(i));
    elseif nodesy(i) >=1000 && nodesy(i) < 10000
        fprintf(fid,'      %d', nodesy(i));
    else
        fprintf(fid,'     %d',nodesy(i));
    end
 
    if linenr == 8
        fprintf(fid,'\n');
        linenr = 0;
    end
end
fprintf(fid,'*END\n');

fclose(fid);

File: LS-OPT_test/createBC.py
#! /usr/bin/env python3

from os.path import exists
from os import remove

##############################################################################
# Basic parameters

readfile = 'lsppout'
writefile = 'bcinfo.k'
#
ytop = 20.0
ybot = -20.0

TOL_g = 1.0e-4

##############################################################################
## Read lsppout file

fid = open(readfile,'r')

flag = False
# print(fid.read())

nodesz = []
nodesy = []
for text in fid:
    #print(x)
    
    if 'NODE' in text:
        flag = True
        
        
    if flag:
        try:
            text = text.split()
            node = int(text[0])
            x    = float(text[1])
            y    = float(text[2])
            z    = float(text[3])
            
            nodesz.append(node)
            
            if y > ytop - TOL_g or y < ybot + TOL_g:
                nodesy.append(node)
        except ValueError:
            print('String cannot be converted')
   

remz = len(nodesz) % 8
remy = len(nodesy) % 8

if remz > 0:     
    nodesz = nodesz + [0]*(8-remz) 

if remy > 0:
    nodesy = nodesy + [0]*(8-remy)

fid.close()

##############################################################################
# Write data

if exists(writefile):
    remove(writefile)
    
fid = open(writefile,'w')

# Write Z list
fid.write('*KEYWORD\n')
fid.write('*SET_NODE_LIST\n')
fid.write('$#     sid       da1       da2       da3       da4    solver\n')
fid.write('         1       0.0       0.0       0.0       0.0MECH\n')
fid.write('$#    nid1      nid2      nid3      nid4      nid5      nid6      nid7      nid8\n')


linenr = 0
for node in range(len(nodesz)):
    linenr += 1
    
    
    if nodesz[node] < 10:
        text = '         {}'.format(nodesz[node])
    elif nodesz[node] >= 10 and nodesz[node] < 100:
        text = '        {}'.format(nodesz[node])
    elif nodesz[node] >= 100 and nodesz[node] < 1000:
        text = '       {}'.format(nodesz[node])
    elif nodesz[node] >= 1000 and nodesz[node] < 10000:
        text = '      {}'.format(nodesz[node])
    else:
        text = '     {}'.format(nodesz[node])
    
    fid.write(text)
    
    if linenr == 8:
        fid.write('\n')
        linenr = 0


# Write Y list
fid.write('*SET_NODE_LIST\n')
fid.write('         2       0.0       0.0       0.0       0.0MECH\n')

linenr = 0
for node in range(len(nodesy)):
    linenr += 1
    
    
    if nodesy[node] < 10:
        text = '         {}'.format(nodesy[node])
    elif nodesy[node] >= 10 and nodesy[node] < 100:
        text = '        {}'.format(nodesz[node])
    elif nodesy[node] >= 100 and nodesy[node] < 1000:
        text = '       {}'.format(nodesy[node])
    elif nodesy[node] >= 1000 and nodesy[node] < 10000:
        text = '      {}'.format(nodesy[node])
    else:
        text = '     {}'.format(nodesy[node])
    
    fid.write(text)
    
    if linenr == 8:
        fid.write('\n')
        linenr = 0    

fid.write('*END\n')
fid.close()

print('--- bcinfo.k file created ---')

